src/Compilers/Core/Portable/CommandLine/ErrorLogger.cs

```diff
@@ -9,6 +9,8 @@ namespace Microsoft.CodeAnalysis
     /// </summary>
     internal abstract class ErrorLogger
     {
+        public Compilation? Compilation { get; set; }
```

Comment:
This feels a bit weird to have a nullable Compilation property, when ErrorLogger will always have a non-null Compilation. I would implement this differently:

1. Revert this change
2. Update the signature of `LogDiagnostic` below to `public abstract void LogDiagnostic(Diagnostic diagnostic, Compilation compilation);`
3. `LogDiagnostic` implementation in `SarifV2ErrorLogger` always logs `Justification` property, no compilation null check to guard the newly added logic.

Reply:
I was thinking the exact same thing. However, `compilation` is useless in `SarifV1ErrorLogger.LogDiagnostic`, to keep the signature of  `SarifV1ErrorLogger.LogDiagnostic` and `SarifV2ErrorLogger.LogDiagnostic` be same, I added the property. Besides, adding a `compilation` argument for `LogDiagnostic` would cause an interdependency of `CommonCompiler.CreateCompilation` and `CommonCompiler.GetErrorLogger`.

Reply:
Never mind. Please take a took at the new commit, which caches a `SuppressionInfo` in the `Diagnostic`.